File: Q2611_「プラス・マイナス・ゼロ」問題/Answer.vb
' Try CodeIQ
' author: Leonardone @ NEETSDKASU
Imports System
 
Module Main
	Sub Main()
	    Dim n%, i%, j%:
	    Dim a&(4000), b&(4000), ans&
	    n = CInt(Console.ReadLine)
	    a(2000) = 1
	    For i = 1 To n
	        If i Mod 2 = 0 Then
	            For j = 0 To 4000
	                If b(j) = 0 Then Continue For
	                a(j + i) += b(j)
	                a(j - i) += b(j)
	                b(j) = 0
	            Next j
	        Else
	            For j = 0 To 4000
	                If a(j) = 0 Then Continue For
	                b(j + i) += a(j)
	                b(j - i) += a(j)
	                a(j) = 0
	            Next j
	        End If
	    Next i
	    ans = IIf(n Mod 2 = 0, a(2000), b(2000))
	    Console.WriteLine(ans)
	End Sub
End Module

File: Q2611_「プラス・マイナス・ゼロ」問題/Answer.R
# Try CodeIQ
# author: Leonardone @ NEETSDKASU
# ==================================================================

solve <- function() {
    k <- 7
    ub <- as.integer(10 ^ k)
    fmt <- paste("%0", toString(k), "d", sep="")
    n <- as.integer(readLines("stdin")[1])
    a <- array(vector("integer", 4000))
    aa <- a
    b <- a
    bb <- b
    fa <- function(x,y) {
        a[x] <<- a[x] + b[y]
        aa[x] <<- aa[x] + bb[y]
        if (a[x] > ub) { aa[x] <<- aa[x] + 1; a[x] <<- a[x] - ub }
    }
    fb <- function(x,y) {
        b[x] <<- b[x] + a[y]
        bb[x] <<- bb[x] + aa[y]
        if (b[x] > ub) { bb[x] <<- bb[x] + 1; b[x] <<- b[x] - ub }
    }
    a[2000] = 1
    for ( i in 1:n )
        if (i %% 2 == 0) 
            for ( j in 1:4000 ) {
                if (b[j] == 0) next
                fa(j + i, j)
                fa(j - i, j)
                b[j] = 0
                bb[j] = 0
            }
        else 
            for ( j in 1:4000 ) {
                if (a[j] == 0) next
                fb(j + i, j)
                fb(j - i, j)
                a[j] = 0
                aa[j] = 0
            }
    pr <- function(x,y) if (x > 0) cat(x, sprintf(fmt, y), sep="") else cat(y)
    if (n %% 2 == 0)
        pr(aa[2000],a[2000])
    else
        pr(bb[2000],b[2000])
    cat("\n")
}

solve()
